src/Math/Lexer.php

```diff
@@ -0,0 +1,324 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Math\Lexer.
+ */
+
+namespace Drupal\rules\Math;
+
+use Drupal\rules\Math\Exception\IncorrectParenthesisException;
+use Drupal\rules\Math\Exception\IncorrectExpressionException;
+use Drupal\rules\Math\Exception\UnknownConstantException;
+use Drupal\rules\Math\Exception\UnknownFunctionException;
+use Drupal\rules\Math\Exception\UnknownOperatorException;
+use Drupal\rules\Math\Exception\UnknownTokenException;
+use Drupal\rules\Math\Token\ParenthesisCloseToken;
+use Drupal\rules\Math\Token\ParenthesisOpenToken;
+use Drupal\rules\Math\Token\CommaToken;
+use Drupal\rules\Math\Token\FunctionToken;
+use Drupal\rules\Math\Token\NumberToken;
+use Drupal\rules\Math\Token\OperatorTokenInterface;
+
+/**
+ * Lexical analyzer for mathematical expressions.
+ */
+class Lexer {
+
+  /**
+   * Static cache of the compiled regular expression.
+   *
+   * @var string
+   */
+  protected $compiledRegex;
+
+  /**
+   * The list of registered operators.
+   *
+   * @var string[]
+   */
+  protected $operators = [];
+
+  /**
+   * The list of registered constants.
+   *
+   * @var array
+   */
+  protected $constants = [];
+
+  /**
+   * The list of registered functions.
+   *
+   * @var array[]
+   */
+  protected $functions = [];
+
+  /**
+   * Registers a function with the lexer.
+   *
+   * @param string $name
+   *   The name of the function.
+   * @param callable $function
+   *   The callback to be executed.
+   * @param int $arguments
+   *   The number of arguments required for the function.
+   *
+   * @return $this
+   */
+  public function addFunction($name, callable $function, $arguments = 1) {
```

Comment:
I would like to avoid reflection due to performance concerns. Do you think it is viable to do this on runtime? Or are you saying that, once I move this to some other place, I would do this during the building of the container?